lib/typecheck.ts

```diff
@@ -392,1195 +421,1293 @@ async function typeCheckFilter(ast : Ast.BooleanExpression,
-            throw new TypeError('Filter required');
-    } else if (ast.table !== undefined && ast.table !== null) {
-        addRequiredInputParamsTable(ast.table, new Set<string>());
-        await typeCheckTable(ast.table, schemas, scope, classes, useMeta);
-        if (ast.table.schema!.require_filter)
-            throw new TypeError('Filter required');
-    }
-    scope.$has_event = !!(ast instanceof Ast.Rule || ast.table);
-
-    if (ast.actions.some((a) => a.isNotify)
-        && (ast instanceof Ast.Command || ast instanceof Ast.OnInputChoice)
-        && !ast.table)
-        throw new TypeError('Cannot return a result without a GET function');
-
-    for (const prim of ast.actions)
-        addRequiredInputParamsAction(prim);
-    await Promise.all(
-        ast.actions.map((action) => typeCheckAction(action, schemas, scope, classes, useMeta)));
-}
+        const scope = new Scope(parentScope);
 
-async function typeCheckProgram(ast : Ast.Program,
-                                schemas : SchemaRetriever,
-                                useMeta = false,
-                                classes : ClassMap = {},
-                                parentScope : Scope|null = null) : Promise<void> {
-    ast.classes.forEach((ast) => {
-        classes[ast.name] = ast;
-    });
-
-    const scope = new Scope(parentScope);
-    if (ast.principal !== null)
-        await typecheckPrincipal(ast.principal, schemas, classes, useMeta);
-
-    for (const klass of ast.classes)
-        await typeCheckClass(klass, schemas, classes, useMeta, false);
-    for (const decl of ast.declarations) {
-        scope.clean();
-        await typeCheckDeclaration(decl, schemas, scope, classes, useMeta);
-    }
-    /*if (ast.rules.length === 0 && ast.oninputs.length === 0)
-        throw new TypeError(`A program must include at least one executable or oninput statement`);*/
-
-    for (const decl of ast.rules) {
-        scope.clean();
-        if (decl instanceof Ast.Assignment)
-            await typeCheckAssignment(decl, schemas, scope, classes, useMeta);
-        else
-            await typeCheckRule(decl, schemas, scope, classes, useMeta);
+        for (const klass of ast.classes)
+            await this.typeCheckClass(klass, false);
+        for (const decl of ast.declarations) {
+            scope.clean();
+            await this._typeCheckDeclaration(decl, scope);
+        }
+
+        for (const decl of ast.statements) {
+            scope.clean();
+            if (decl instanceof Ast.Assignment)
+                await this._typeCheckAssignment(decl, scope);
+            else
+                await this._typeCheckExpressionStatement(decl, scope);
+        }
     }
-    for (const choice of ast.oninputs) {
-        scope.clean();
-        await typeCheckRule(choice, schemas, scope, classes, useMeta);
+
+    async typeCheckProgram(ast : Ast.Program,
+                           parentScope : Scope|null = null) : Promise<void> {
```

Comment:
why is this function of type `Promise<void>`?  there are a few other cases where nothing is returned, but the function is annotated with `Promise<void>`; there are also a few other cases that are the other way around.

Reply:
Async functions that return nothing return `Promise<void>`, and regular functions that return nothing are `void`. In both cases the return type annotation can usually be omitted and typescript infers it, except for publicly exported functions and public methods of exported classes we used to have a eslint warning if the function lacks a type annotation. I recently silenced it because I got tired of typing, but yeah.